File: packages/owl_ads_gromore/android/src/main/kotlin/com/owlllwo/plugins/gromore/InitializationCoordinatorState.kt
package com.owlllwo.plugins.gromore

/** Pure process-initialization state. It never calls or simulates an ad SDK. */
internal class InitializationCoordinatorState<Owner : Any, Fingerprint : Any> {
    internal enum class Phase { IDLE, STARTING, STARTED, STALLED }

    internal sealed interface RequestDecision {
        data class Start(val attemptId: Long) : RequestDecision

        data class Wait(val attemptId: Long) : RequestDecision

        data object Complete : RequestDecision

        data object Conflict : RequestDecision

        data object RestartRequired : RequestDecision
    }

    var phase: Phase = Phase.IDLE
        private set

    private var fingerprint: Fingerprint? = null
    private var attemptId = 0L
    private val owners = linkedSetOf<Owner>()

    fun request(
        owner: Owner,
        requestedFingerprint: Fingerprint,
        sdkReady: Boolean,
    ): RequestDecision = when (phase) {
        Phase.IDLE -> {
            fingerprint = requestedFingerprint
            attemptId += 1
            owners += owner
            phase = Phase.STARTING
            RequestDecision.Start(attemptId)
        }

        Phase.STARTING -> {
            if (fingerprint != requestedFingerprint) {
                RequestDecision.Conflict
            } else {
                owners += owner
                RequestDecision.Wait(attemptId)
            }
        }

        Phase.STARTED -> when {
            fingerprint != requestedFingerprint -> RequestDecision.Conflict
            sdkReady -> RequestDecision.Complete
            else -> RequestDecision.RestartRequired
        }

        Phase.STALLED -> when {
            fingerprint != requestedFingerprint -> RequestDecision.Conflict
            sdkReady -> {
                phase = Phase.STARTED
                RequestDecision.Complete
            }

            else -> RequestDecision.RestartRequired
        }
    }

    /** Removes a detached/disposed owner without changing the SDK attempt. */
    fun remove(owner: Owner): Boolean = owners.remove(owner)

    /** Expires one owner. The process attempt becomes stalled after its last timeout. */
    fun timeout(owner: Owner, expectedAttemptId: Long): Boolean {
        if (phase != Phase.STARTING || attemptId != expectedAttemptId || !owners.remove(owner)) {
            return false
        }
        if (owners.isEmpty()) phase = Phase.STALLED
        return true
    }

    /** Accepts only the current native success and returns the live owners to complete. */
    fun nativeSucceeded(expectedAttemptId: Long): Set<Owner>? {
        if ((phase != Phase.STARTING && phase != Phase.STALLED) || attemptId != expectedAttemptId) {
            return null
        }
        val completedOwners = owners.toSet()
        owners.clear()
        phase = Phase.STARTED
        return completedOwners
    }

    /** Accepts only the current native failure, returning to a retryable idle state. */
    fun nativeFailed(expectedAttemptId: Long): Set<Owner>? {
        if ((phase != Phase.STARTING && phase != Phase.STALLED) || attemptId != expectedAttemptId) {
            return null
        }
        val completedOwners = owners.toSet()
        owners.clear()
        fingerprint = null
        phase = Phase.IDLE
        return completedOwners
    }
}

File: packages/owl_ads_gromore/ios/owl_ads_gromore/Sources/owl_ads_gromore/InitializationCoordinatorState.swift
import Foundation

/// Pure process-initialization state. It never calls or simulates an ad SDK.
final class InitializationCoordinatorState<Owner: Hashable, Fingerprint: Equatable> {
  enum Phase: Equatable {
    case idle
    case starting
    case started
    case stalled
  }

  enum RequestDecision: Equatable {
    case start(attemptID: UInt64)
    case wait(attemptID: UInt64)
    case complete
    case conflict
    case restartRequired
  }

  private(set) var phase = Phase.idle
  private var fingerprint: Fingerprint?
  private var attemptID: UInt64 = 0
  private var owners: Set<Owner> = []

  func request(
    owner: Owner,
    fingerprint requestedFingerprint: Fingerprint,
    sdkReady: Bool
  ) -> RequestDecision {
    switch phase {
    case .idle:
      fingerprint = requestedFingerprint
      attemptID += 1
      owners.insert(owner)
      phase = .starting
      return .start(attemptID: attemptID)
    case .starting:
      guard fingerprint == requestedFingerprint else { return .conflict }
      owners.insert(owner)
      return .wait(attemptID: attemptID)
    case .started:
      guard fingerprint == requestedFingerprint else { return .conflict }
      return sdkReady ? .complete : .restartRequired
    case .stalled:
      guard fingerprint == requestedFingerprint else { return .conflict }
      guard sdkReady else { return .restartRequired }
      phase = .started
      return .complete
    }
  }

  /// Removes a detached/disposed owner without changing the SDK attempt.
  @discardableResult
  func remove(owner: Owner) -> Bool {
    owners.remove(owner) != nil
  }

  /// Expires one owner. The process attempt becomes stalled after its last timeout.
  @discardableResult
  func timeout(owner: Owner, attemptID expectedAttemptID: UInt64) -> Bool {
    guard phase == .starting, attemptID == expectedAttemptID,
      owners.remove(owner) != nil
    else { return false }
    if owners.isEmpty { phase = .stalled }
    return true
  }

  /// Accepts only the current native success and returns the live owners to complete.
  func nativeSucceeded(attemptID expectedAttemptID: UInt64) -> Set<Owner>? {
    guard (phase == .starting || phase == .stalled), attemptID == expectedAttemptID else {
      return nil
    }
    let completedOwners = owners
    owners.removeAll()
    phase = .started
    return completedOwners
  }

  /// Accepts only the current native failure, returning to a retryable idle state.
  func nativeFailed(attemptID expectedAttemptID: UInt64) -> Set<Owner>? {
    guard (phase == .starting || phase == .stalled), attemptID == expectedAttemptID else {
      return nil
    }
    let completedOwners = owners
    owners.removeAll()
    fingerprint = nil
    phase = .idle
    return completedOwners
  }
}
